feat: support tags/hashtags on examples to improve discovery #422

srini-x opened the pull request:
Closes #422 

1. Added a new page `Tags` between `Gallery` and `Hello World!`
2. The `tags` are subheadings on this page, sorted alphabetically. Under each `tag`, links to the examples with those tags are available.
3. Added a `Tags` field under each example with tags. These tags are links that link back to the Tags page mentioned above.
4. Added `keywords` metadata in the header of the each generated `.md` file. Docusaurus [markdown headers](https://v2.docusaurus.io/docs/markdown-features#markdown-headers) provide only `keywords` under Docs and `tags` under Blog.
5. In the python example files, tags can be provided as the last line of description.
6. Tags line can not have `.` in it. Tags must be `,` separated. A tag can not contain spaces.

ToDo:
1. Comment the code
2. Add tags to all remaining examples.

@lo5, If you like this implementation, I will go ahead and complete the ToDos listed above.

example:

`buttons.py`

```py
# Form / Buttons
# Use the `ui.buttons()` function to group related buttons.
# form, buttons
# ---
from h2o_wave import main, app, Q, ui
```

`buttons.md`

```md
---
title: Form / Buttons
keywords:
  - form
  - buttons
---
```

![wave_docs_examples_tags](https://user-images.githubusercontent.com/16831326/103091836-476cc680-45aa-11eb-8634-3c160ba766b9.png)

![wave_docs_examples_buttons_tags](https://user-images.githubusercontent.com/16831326/103091843-4cca1100-45aa-11eb-85f6-427a9b605474.png)

lo5 reviewed (changes requested):
@srini-x This is a great idea - thanks for the PR.

It would be good for `get_keywords_if_available()` to be a `parse_tags(description: str)` that accepts a description containing tags and returns a `(new_description, tags)` tuple.

The convention for tags can be `#tag`. For example, accept a description 
```
Use a checklist to group a set of related checkboxes. #form #checkbox #checklist
```
or
```
Use a checklist to group a set of related checkboxes. 
#form #checkbox #checklist
```

or

```
Use a #checklist to group a set of related checkboxes. 
#form #checkbox
```

... and return `('Use a checklist to group a set of related checkboxes.', ['form', 'checkbox', 'checklist'])`

This way, tags can be optional, free-form, and we don't have to stick to a convention that the last line is reserved for tags, which might seem odd when one is staring at the example's header in isolation in an editor.

lo5 commented:
@srini-x - Also, this has minor conflicts with @dulajra 's change. Let's name the new page "Examples by Tag" and add it after the "All Examples" page.

srini-x commented:
@lo5

- implemented `parse_tags` function
- added tags for all examples
- double checked for typos and broken tags

![wave_examples_by_tag](https://user-images.githubusercontent.com/16831326/103149476-9f2c3e80-471e-11eb-91bd-27550678241d.png)

lo5 commented on py/sync_examples.py:
```diff
@@ -62,15 +91,82 @@ def write_file(p: str, txt: str) -> str:
+    # Remove the '#' before the tags in description
+    new_d = pattern.sub(r'\1\2', ' ' + description)
+
+    # Remove the last line in description if it has only tags
```
Not sure why this part is required. We could simply trim the description.

srini-x replied:
@lo5 This is to make this example work.

```
Use a #checklist to group a set of related checkboxes. 
#form #checkbox
```

If the tags are part of the description, we don't want to remove the entire word, but only the `#` in front of it.

I trim the description, if the tags are at the end of it.

lo5 commented on py/sync_examples.py:
```diff
@@ -98,6 +194,30 @@ def make_gallery(examples: List[Example]):
+    for e in examples:
+        for t in e.tags:
+            tags_dict[t].append(e)
+    return sorted(tags_dict.items())
```
Need to sort by the example's title.

srini-x replied:
@lo5 `label` is for `ui.label` and `labels` is for data labels in plots.

lo5 replied:
Yes, in that case let's drop `labels`. Thanks @srini-x.

lo5 reviewed (changes requested):
Thanks for the PR @srini-x.

The following tags look overly granular to me, and can be removed - most of these have one example under them, which makes the tags page less useful:
- application
- area
- async
- axis_title
- backend
- bar
- bubble_plot
- clock
- color
- column
- combined
- curve
- custom
- d3
- data
- dynamic
- fill
- filter
- font_size
- game_of_life
- gantt
- gauge
- glider_gun
- global_variables
- group_by
- grouped
- heatmap
- helix
- hilbert
- issue_tracker
- large
- layer
- line
- live
- multi_user
- multiple_categories
- occlusion
- opacity
- path
- point
- polar
- polygon
- position
- primitives
- racetrack
- realtime
- region
- rose
- scatterplot
- shape
- simple
- size
- small
- smooth
- sort
- spline
- state
- status
- step
- step_after
- step_before
- stroke
- stroked
- tall
- theta
- title
- transpose
- turtle
- update
- wide

Additionally:
- labels -> merge plural with singular
- preselection -> merge selection
- vega_lite -> merge with vega

srini-x commented:
@lo5 

fixed the following

- [x] Removed the overly granular tags
- [x] Merged `labels` -> `label`
- [x] Merged `preselection` -> `selection`
- [x] Merged `vega_lite` -> `vega`
- [x] Merged `nav` -> `navigation`
- [x] Examples within each tag are sorted

srini-x commented:
@lo5 merged commits from master

lo5 commented on py/examples/table.py:
```diff
@@ -1,9 +1,11 @@
 from faker import Faker
-from h2o_wave import main, app, Q, ui
+
+from h2o_wave import app, Q, ui
```
The `main` import is required here.

srini-x replied:
fixed.

srini-x commented:
@lo5 removed `labels` tag.

lo5 reviewed (changes requested):
Thanks @srini-x - looks good. We'll merge after the correction to `py/examples/table.py`.

lo5 reviewed (approved):
Thanks @srini-x!
